### ISSUE TITLE
Negative numbers in types like-uint

### ISSUE THREAD

**Issue by kless** [rust-lang/rust#5477]
The compiler should show a warning message when is used a negative number into a like-uint type

```rust
let n: u8 = -1;
println!("{}", n); // => 255
```

**Comment by mbrubeck** [comment]
PistonDevelopers/image#305 is an example of a real-world bug that would have been caught by this warning.  That bug involved code like `std::iter::range_step(len, 0, -1)` where `len` is unsigned.

**Comment by nagisa** [comment]
[Relevant internals discussion](http://internals.rust-lang.org/t/forbid-unsigned-integer/752)

**Comment by pnkfelix** [comment]
see #23945; we're going further than a lint and moving to a feature-gate at this point.

The goals are two-fold:
 1. Removing unary-negation on unsigned integers does seem more consistent with our current overflow story, and
 2. if the community *does* strongly want `-expr` to work on unsigned integers, then this is a way to force the community to tell us.  (Perhaps using stronger language than we might otherwise want, we shall see...)

**Comment by SiegeLord** [comment]
Code like this compiles and runs without error in stable rustc today:

```
let a = -1; // A gets inferred to u32, but the code compiles without error
let b: u32 = a;
```

Is the intention to stop that code from compiling or issue a runtime error?

**Comment by nagisa** [comment]
Yes, although that would be somewhat of a breaking change.

**Comment by pnkfelix** [comment]
The intention was to force examples like [above comment](https://github.com/rust-lang/rust/issues/5477#issuecomment-120824098) to be feature-gated (and thus you'd have to opt-in to be able to write them).  But it seems like the feature-gating code in #23945 did not catch this case (or it has regressed since then, which seems less likely to me...)

**Comment by rnestler** [comment]
@SiegeLord 
```Rust
let a = -1; // 'a' gets inferred to u32, but the code compiles without error
let b: u32 = a;
```
Gives a warning in stable and fails to compile in nightly for me on [play.rust-lang.org](https://play.rust-lang.org/?code=fn%20main()%20%7B%0A%20%20%20%20let%20a%20%3D%20-1%3B%20%2F%2F%20%27a%27%20gets%20inferred%20to%20u32%2C%20but%20the%20code%20compiles%20without%20error%0A%20%20%20%20let%20b%3A%20u32%20%3D%20a%3B%0A%20%20%20%20println!(%22a%20is%20%7B%3Ax%7D%2C%20b%20is%20%7B%3Ax%7D%22%2C%20a%2C%20b)%3B%0A%7D&version=nightly)

**Comment by steveklabnik** [comment]
Yes, I would consider this fixed. The example is now fixed, it's behind a feature gate, and so when/if the gate is stabilized or removed, this will end up being discussed, so I think all of the concerns in this ticket are taken care of. Let me know if that's wrong!
